### PR TITLE
service/storage_proxy: Add metrics keeping track of incoming hints

### PR DESCRIPTION
**Pull request by dawmd**
Although Scylla already exposes metrics keeping track of various information related to hinted handoff, all of them correspond to either storing or sending hints. However, when debugging, it's also crucial to be aware of how many hints are coming to a given node and what their size is. Unfortunately, the existing metrics are not enough to obtain that information.

This PR introduces the following new metrics:

* `sent_bytes_total` – the total size of the hints that have been sent from a given shard,
* `received_hints_total` – the total number of hints that a given shard has received,
* `received_hints_bytes_total` – the total size of the hints a given shard has received.

It also renames `hints_manager_sent` to `hints_manager_sent_total` to avoid conflicts of prefixes between that metric and `sent_bytes_total` in tests.

Fixes scylladb/scylladb#10987

### CODE REVIEW

**Comment by dawmd** [comment]
The failing `test_fence_hints` is irrelevant – the test is skipped. The failure in `test_hintedhandoff_switch_config_in_runtime_via_http_api` is most likely caused by scylladb/scylladb#19006.

File: service/storage_proxy.cc
Comment (piodul) [inline service/storage_proxy.cc]: Do we need `.set_skip_when_empty(false)` here? I suppose that the metric will not be larger than 0 on other scheduling groups than `streaming`, so this is a worthwhile optimization. On the other hand, if there are some hints received in other scheduling groups then it will be a clear indication of a bug.

File: service/storage_proxy_stats.hh
Comment (piodul) [inline service/storage_proxy_stats.hh:196]: Why `size_t` and not `uint64_t` like other metrics in this struct?

File: service/storage_proxy.cc
Comment (piodul) [inline service/storage_proxy.cc:2798]: Ah, you removed `.set_skip_when_empty(false)`, but you did it in the wrong commit.

**Review by piodul — COMMENTED** [review]
Left some nitpicks, otherwise LGTM.

File: db/hints/manager.cc
Comment (amnonh) [inline db/hints/manager.cc:177]: Assuming that it's the total bytes ever sent, it should be counter (ever growing) and a better name would be sent_bytes_total

File: service/storage_proxy.cc
Comment (amnonh) [inline service/storage_proxy.cc:2792]: we have deprecated collectd, use counter or gauge.
and a better name will be received_hints_total

File: service/storage_proxy.cc
Comment (amnonh) [inline service/storage_proxy.cc:2796]: use counter,  a better name will be received_hints_bytes_total

**Review by amnonh — CHANGES_REQUESTED** [review]

**Comment by dawmd** [comment]
Version 2–4:
* rebased on top of the current master
* renamed `sent_hints_size` to `sent_hints_bytes_total`, `sent_total_size` to `sent_bytes_total`, `received_hint_total_size` to `received_hints_bytes_total`
* included information that incoming hints are both regular hints and MV hints in the description of the metrics in storage proxy
* started using counter instead of gauge/total_bytes

File: db/hints/manager.cc
Comment (amnonh) [inline db/hints/manager.cc:177]: Why do the receive hints have skip_when_empty, and they sent don't?

File: service/storage_proxy.cc
Comment (amnonh) [inline service/storage_proxy.cc:2814]: received_hints_total

File: db/hints/manager.cc
Comment (dawmd) [reply db/hints/manager.cc:177]: That's a good point. None of the hinted handoff metrics use `skip_when_empty`. I'm not knowledgeable in how exactly it's used (please, correct me if I'm wrong), but the way I understand it is the metrics won't be "displayed" unless they've been modified at least once.

In the context of hints, I think there's a merit in keeping them that way – when analysing the behaviour of a node, hints often play a significant role in how performant it is, so seeing no hints have been sent or written is useful because it gives you that information right away. The same reasoning could be applied to the metrics with incoming hints (which would mean I shouldn't have made them skipped when empty). But I'm sure it's still possible to obtain those metrics, so perhaps it'd be better to skip them when empty anyway. I definitely agree the metrics corresponding to errors, etc. should be skipped.

**Comment by dawmd** [comment]
Version 5:
* `received_hint_count` renamed to `received_hints_total`

**Review by amnonh — APPROVED** [review]
LGTM (just update the message to reflect the updated names)

**Comment by dawmd** [comment]
The PR message updated

**Comment by dawmd** [comment]
Version 6:
* fixed the failing test: `topology_experimental_raft/test_fencing` (`test_fence_hints`)
* updated the commit message accordingly

**Comment by dawmd** [comment]
The failing test is scylladb/scylladb#19006. These changes should have no impact on it

**Comment by mykaul** [comment]
@dawmd - some maintainers ask to rerun CI. Please do.

**Comment by piodul** [comment]
The test `test_hintedhandoff_switch_config_in_runtime_via_http_api` is failing too frequently. There is no use in fighting CI, we should fix it before merging the PR.

**Comment by piodul** [comment]
> The test `test_hintedhandoff_switch_config_in_runtime_via_http_api` is failing too frequently. There is no use in fighting CI, we should fix it before merging the PR.

@dawmd It turns out that the test failure is due to the change in the name of the metric. `test_hintedhandoff_switch_config_in_runtime_via_http_api` is sensitive to the metric name and is also affected by the choice of the new names. Please adjust the test first so that it will work with the old name and the new name of the metric, then we can continue with this PR.

**Comment by piodul** [comment]
@dawmd why did you choose the `backport/6.0` label specifically?

**Review by avikivity — APPROVED** [review]

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
Queued

**Pull request merged**
